### PR TITLE
feat: display only start time when start and end time are the same

### PR DESCRIPTION
**Pull request by Xstoudi**
## Checklist

Please put "X" in the below checkboxes that apply:

- [X] If committing a change of production code, I have tested it in different browsers (Chrome, Firefox, Safari). 
- [X] If committing a new feature, I have first submitted an issue (Please note: you are free to open PRs for non-issued features, but opening an issue increases your chance of a successful PR). 
- [X] If committing a new feature, I have also written the appropriate tests for it. 
- [X] I have tried to build the feature in a way, that it does not cause any breaking changes for others (unless 
  agreed upon in an issue). 

**Premium**
- [ ] I'm a Schedule-X premium user

## This PR solves the following problem**. 

Allow to add ponctual event (start and end are the same) in calendar as proposed in #783.

## How to test this PR**. 

For example:  
1. Simply add an event with start and end that are the same.
2. Or check 2024-10-20 at 4PM on `npm run dev` page

### CODE REVIEW

File: packages/shared/src/utils/stateless/calendar/calendar-event/calendar-event.impl.ts [hunk 1]
```diff
@@ -72,6 +73,7 @@ export default class CalendarEventImpl implements CalendarEventInternal {
 
   get _isSingleHybridDayTimed(): boolean {
     if (!this._config.isHybridDay) return false
+    if (this.start === this.end) return false
```
Comment (tomosterlund) [inline packages/shared/src/utils/stateless/calendar/calendar-event/calendar-event.impl.ts:76]: why do we need this line?

I believe this getter can never be true, if two timed events are identical

File: packages/shared/src/utils/stateless/calendar/calendar-event/calendar-event.impl.ts [hunk 2]
```diff
@@ -100,6 +102,14 @@ export default class CalendarEventImpl implements CalendarEventInternal {
     )
   }
 
+  get _isSingleTime(): boolean {
```
Comment (tomosterlund) [inline packages/shared/src/utils/stateless/calendar/calendar-event/calendar-event.impl.ts:105]: I'd prefer `_isSingleMinute`

File: packages/shared/src/utils/stateless/time/date-time-localization/__test__/get-time-stamp.spec.ts [hunk 3]
```diff
@@ -102,4 +102,28 @@ describe('Getting the time stamp for the event modal', () => {
       expect(result).toEqual(expected)
     })
   })
+
+  describe('Time stamps for ponctual events', () => {
```
Comment (tomosterlund) [inline packages/shared/src/utils/stateless/time/date-time-localization/__test__/get-time-stamp.spec.ts:106]:
```suggestion
  describe('Time stamps for single minute events', () => {
```

File: packages/shared/src/utils/stateless/time/date-time-localization/__test__/get-time-stamp.spec.ts [hunk 4]
```diff
@@ -102,4 +102,28 @@ describe('Getting the time stamp for the event modal', () => {
...
+    const singleTimeEvent: Partial<CalendarEventInternal> = {
```
Comment (tomosterlund) [inline packages/shared/src/utils/stateless/time/date-time-localization/__test__/get-time-stamp.spec.ts:107]:
```suggestion
    const singleMinuteEvent: Partial<CalendarEventInternal> = {
```

File: packages/shared/src/utils/stateless/time/date-time-localization/__test__/get-time-stamp.spec.ts [hunk 5]
```diff
@@ -102,4 +102,28 @@ describe('Getting the time stamp for the event modal', () => {
...
+      start: '2023-10-06 20:07',
+      end: '2023-10-06 20:07',
+      _isSingleTime: true,
+    }
+
+    it('should return a German time stamp for a ponctual event', () => {
```
Comment (tomosterlund) [inline packages/shared/src/utils/stateless/time/date-time-localization/__test__/get-time-stamp.spec.ts:113]:
```suggestion
    it('should return a German time stamp for a single minute event', () => {
```

File: packages/calendar/src/utils/stateless/events/event-styles.ts [hunk 6]
```diff
@@ -10,6 +13,23 @@ export const getEventHeight = (
   dayBoundaries: DayBoundariesInternal,
   pointsPerDay: number
 ) => {
+  if (start === end) {
+    return (
+      timePointToPercentage(
+        pointsPerDay,
+        dayBoundaries,
+        timePointsFromString(
+          timeFromDateTime(addTimePointsToDateTime(end, 100))
```
Comment (tomosterlund) [inline packages/calendar/src/utils/stateless/events/event-styles.ts:22]: 100 equals 1 hours here. Not saying it's wrong, but I'm curious why you think 1 hour is the correct way to display a one-minute event. I personally would have picked the lowest possible number here to be able to display any info at all, like the event title.

Google calendar seems to do something in between our both approaches: a half-hour in this case. What do you think?

File: packages/calendar/src/utils/stateless/events/sort-events-for-week.ts [hunk 7]
```diff
@@ -7,7 +7,11 @@ export const sortEventsForWeekView = (
   const timeGridEvents: CalendarEventInternal[] = []
 
   for (const event of allCalendarEvents) {
-    if (event._isSingleDayTimed || event._isSingleHybridDayTimed) {
+    if (
+      event._isSingleDayTimed ||
+      event._isSingleHybridDayTimed ||
+      event._isSingleTime
```
Comment (tomosterlund) [inline packages/calendar/src/utils/stateless/events/sort-events-for-week.ts:13]: Are you sure this is needed? My first thought is that all `_isSingleTime` events would already be caught by this condition

File: packages/shared/src/utils/stateless/calendar/calendar-event/calendar-event.impl.ts [hunk 8]
```diff
@@ -42,7 +42,8 @@ export default class CalendarEventImpl implements CalendarEventInternal {
     return (
       dateTimeStringRegex.test(this.start) &&
       dateTimeStringRegex.test(this.end) &&
-      dateFromDateTime(this.start) === dateFromDateTime(this.end)
+      dateFromDateTime(this.start) === dateFromDateTime(this.end) &&
```
Comment (tomosterlund) [inline packages/shared/src/utils/stateless/calendar/calendar-event/calendar-event.impl.ts:45]: I think we should not make this change. Semantically it makes more sense if an event can be `_isSingleDayTimed`  _and_ `_isSingleMinute`. Then we can also undo the change in `packages/calendar/src/utils/stateless/events/sort-events-for-week.ts`

File: packages/calendar/src/components/week-grid/time-grid-event.tsx [hunk 9]
```diff
@@ -59,6 +59,11 @@ export default function TimeGridEvent({
     const localizedStartTime = toJSDate(start).toLocaleTimeString(
       ...localizeArgs
     )
+
+    if (start === end) {
```
Comment (tomosterlund) [inline packages/calendar/src/components/week-grid/time-grid-event.tsx:63]: I think we should also have a unit test for this.

**Review by tomosterlund — CHANGES_REQUESTED** [review]

**Comment by tomosterlund** [comment]
I like your approach as a whole: I think you found the right way to do this :)

suggested a few improvements, and one thing more:

I think it would make sense to also change the way these events appear in the event modal. Do you mind updating the event modal too?

**Comment by tomosterlund** [comment]
I deleted a bunch of comments, because I changed my mind a little on how we should best implement this.

I think making changes in calendar-event.impl adds a bit too much complexity. The **event itself** doesn't need to change for this. I think we should just change the way it is displayed. Probably it's enough checking if `start === end` for timed events, and when that's the case, only display the localized start

File: packages/calendar/src/utils/stateless/events/event-styles.ts (hunk 6 above)
Comment (Xstoudi) [reply packages/calendar/src/utils/stateless/events/event-styles.ts:22]: To be honest I didn't have an opinion on how to show it on other type of calendar than monthly. Google Calendar approach seems like a good compromise between showing it big enough but not too big tho.

File: packages/calendar/src/components/week-grid/time-grid-event.tsx (hunk 9 above)
Comment (Xstoudi) [reply packages/calendar/src/components/week-grid/time-grid-event.tsx:63]: Done, let me know if it's not enough

**Comment by Xstoudi** [comment]
Hey @tomosterlund,

I fixed your comments, let me know if anything else is missing!

**Review by tomosterlund — APPROVED** [review]
Looks better now.

Thanks for taking the time to contribute :)
